added sample_filter_outputs utility and accompanying simple tests

Dekermanjian opened the pull request:
This addresses #521, allows state space users to sample filtered|predicted|smoothed|observed states|covariances using a utility that is consistent with PyMC workflow of sample -> sample_posterior.

jessegrabowski commented on pymc_extras/statespace/core/statespace.py:
```diff
@@ -1678,6 +1678,91 @@ def sample_statespace_matrices(
+        compile_kwargs.setdefault("mode", self.mode)
+
+        with pm.Model(coords=self.coords) as m:
+            pm_mod = modelcontext(None)
```
This will always return `m` (the current model context). What is `pm_mod` used for?

Dekermanjian replied:
Yeah, this was wrong. I removed it.

jessegrabowski reviewed (changes requested):
This looks great to me, just a small question about the use of `modecontext` here

jessegrabowski commented:
Also you need to rebase :)

jessegrabowski commented on pymc_extras/statespace/core/statespace.py:
```diff
@@ -1678,6 +1678,91 @@ def sample_statespace_matrices(
+
+            for output in filter_output_names:
+                match output.name:
+                    case "filtered_states" | "predicted_states" | "smoothed_states":
+                        dims = [TIME_DIM, "state"]
+                    case "filtered_covariances" | "predicted_covariances" | "smoothed_covariances":
+                        dims = [TIME_DIM, "state", "state_aux"]
+                    case "observed_states":
+                        dims = [TIME_DIM, "observed_state"]
+                    case "observed_covariances":
```
A dictionary mapping cases to dims might be better than the casework here. There isn't already one in constants?

Dekermanjian replied:
There was a dictionary in `constats.py` variable name `FILTER_OUTPUT_DIMS` but the key names are singular whereas the returned names from both `kalman_(filter|smoother).build_graph()` are plural. I handled this inside the method but I am wondering if you want to consolidate this either in `constants.py` of inside the `kalman_(filter|smoother).build_graph()` methods?

Dekermanjian commented on pymc_extras/statespace/core/statespace.py:
```diff
@@ -1678,6 +1678,91 @@ def sample_statespace_matrices(
 
+    def sample_filter_outputs(
+        self, idata, filter_output_names: str | list[str] | None, group: str = "posterior", **kwargs
+    ):
```
I added this check in.

Dekermanjian commented:
Hey @jessegrabowski, I updated some of the constant in `constants.py` to be plural and updated the tests for any mismatches. There were a few constants that I was not sure about so I wanted to ask you first before I change them. These are:

```
ALL_STATE_DIM = "state(?s)"
ALL_STATE_AUX_DIM = "state(?s)_aux"
OBS_STATE_DIM = "observed_state(?s)"
OBS_STATE_AUX_DIM = "observed_state(?s)_aux"

NEVER_TIME_VARYING = ["initial_state(?s)", "initial_state(?s)_cov(?s)"]
VECTOR_VALUED = ["initial_state(?s)", "state(?s)_intercept(?s)", "obs_intercept(?s)"]

LONG_MATRIX_NAMES = [
    "initial_state(?s)",
    "initial_state(?s)_cov(?s)",
    "state(?s)_intercept(?s)",
    "obs_intercept(?s)",
    "obs_cov(?s)",
    "state_cov(?s)",
]
```

jessegrabowski commented:
Dims should be singular, I have strong feelings on that.

For the matrix names, I have less of a strong preference. On one hand, `x0` is a vector of states, but on the other hand, it is *the* state vector. So it could go either way. I guess I lean to not changing things in that case?

If you agree, that would mean all of the things you identified there would stay as-is I think.

Dekermanjian commented:
I agree that the singular dim names sound better. No objections from me about keeping the rest as-is! I think the main issue of having the same thing named differently is now resolved 🤞

jessegrabowski reviewed (changes requested):
A few final nitpicks then let's merge this! It's looking really great.

Dekermanjian commented on pymc_extras/statespace/core/statespace.py:
```diff
@@ -1678,6 +1678,93 @@ def sample_statespace_matrices(
+                T, R, Q, filter_outputs[0], filter_outputs[3]
+            )
+
+            # Filter output names are singular in constants.py but are returned as plural from kalman_.build_graph()
+            # filter_output_dims_mapping = {}
+            # for k in FILTER_OUTPUT_DIMS.keys():
+            #     filter_output_dims_mapping[k + "s"] = FILTER_OUTPUT_DIMS[k]
```
oops! Sorry about this. That was a careless oversight. I will clean that up right away!

jessegrabowski commented on pymc_extras/statespace/core/statespace.py:
```diff
@@ -1684,6 +1684,21 @@ def sample_filter_outputs(
         if isinstance(filter_output_names, str):
             filter_output_names = [filter_output_names]
 
+        drop_keys = {"predicted_observed_states", "predicted_observed_covariances"}
```
I think we shouldn't treat these as special (even though I agree it's silly to ask for them). I'd be confused if I tried to ask for them and it said it's not a valid filter output name.

Having everything in one place is convenient, even if it's duplicative.

Dekermanjian replied:
@jessegrabowski, I believe this happens in `_postprocess_scan_results()` in `kalman_filter.py`. It looks like the names of the filter outputs are hardcoded in there.

jessegrabowski replied:
If you want to make this consistent in this PR I have no objection. I don't have a good sense if you should change FILTER_OUTPUT_DIMS to match the output names, or change the output names to match the FILTER_OUTPUT_DIMS. I'll defer to you if you have a sense of which one is better.

Dekermanjian commented:
> I made one last nitpick, but it's not a blocker. Feel free to address or not, then merge :)

Hey @jessegrabowski, I don't believe I have permissions to merge.

jessegrabowski commented:
Great work as always :D

Dekermanjian commented:
Thank you, Jesse! Always happy to help!
